File: Data Structures & Algorithms/three-integer-sum/submission-5.java
class Solution {
  public List<List<Integer>> threeSum(int[] nums) {

    Arrays.sort(nums);
    List<List<Integer>> res = new ArrayList<>();
    for (int i = 0; i < nums.length - 2; i++) {
      if (i > 0 && nums[i] == nums[i - 1]) {
        continue;
      }

      int lPtr = i + 1;
      int rPtr = nums.length - 1;

      while (lPtr < rPtr) {
        int num1 = nums[i];
        int num2 = nums[lPtr];
        int num3 = nums[rPtr];

        int sum = num1 + num2 + num3;

        if (sum > 0) {
          rPtr--;
        } else if (sum < 0) {
          lPtr++;
        } else {
          res.add(List.of(num1, num2, num3));
          lPtr++;
          rPtr--;
          while (nums[lPtr] == nums[lPtr - 1] && lPtr < rPtr) lPtr++;
        }
      }
    }
    return res;
  }

}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-5.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        return self.findMinPtr(nums, 0, len(nums) - 1)

    def findMinPtr(self, nums, lptr, rptr) -> int:
        
        if lptr == rptr:
            return nums[lptr]

        mid = (rptr + lptr) // 2
        
        if nums[mid] > nums[-1]:
            return self.findMinPtr(nums, mid+1, rptr)
        else:
            return self.findMinPtr(nums, lptr, mid)
